graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/commonjs/NpmCompatibleESModuleLoader.java

```diff
@@ -98,9 +101,31 @@ public final class NpmCompatibleESModuleLoader extends DefaultESModuleLoader {
     private static final String INVALID_MODULE_SPECIFIER = "Invalid module specifier: '";
     private static final String UNSUPPORTED_FILE_EXTENSION = "Unsupported file extension: '";
     private static final String UNSUPPORTED_PACKAGE_EXPORTS = "Unsupported package exports: '";
+    private static final String INVALID_PACKAGE_EXPORT = "Invalid package export: ";
     private static final String UNSUPPORTED_PACKAGE_IMPORTS = "Unsupported package imports: '";
     private static final String UNSUPPORTED_DIRECTORY_IMPORT = "Unsupported directory import: '";
     private static final String INVALID_PACKAGE_CONFIGURATION = "Invalid package configuration: '";
+    private static final String EXPORT_TYPE_GRAALJS = "graaljs";
+    private static final String EXPORT_TYPE_IMPORT = "import";
+    private static final String EXPORT_TYPE_REQUIRE = "require";
+    private static final String EXPORT_TYPE_DEFAULT = "default";
+    private static final LinkedList<String> EXPORT_TYPES;
+
+    static {
+        EXPORT_TYPES = new LinkedList<>(
+                List.of(EXPORT_TYPE_GRAALJS, EXPORT_TYPE_IMPORT, EXPORT_TYPE_REQUIRE, EXPORT_TYPE_DEFAULT)
+        );
+    }
+
+    public static void registerPreferredExportType(String exportType) {
```

Comment:
Where is this used and for what?

Reply:
This would be used by embedders:
```java
NpmCompatibleESModuleLoader.registerPreferredExportType("elide");
```
... would register the engine `elide` as a preferred token, greater in precedence than `graaljs`, `import`, `require`, or `default`. Perhaps a browser builder could register `"browser"` to prefer imports of that type.

Reply:
I see... could you do this in a way that does not require modifying a `static final` list?

Reply:
I can try; for example, what if the embedder registers a callback to provide a list of candidate strings? I just figure the callback would probably not need to be dynamic in nature (these strings are likely registered once at startup and not again).

I could wrap the list in an atomic, but that wouldn't do much. What would you recommend as an approach for this one?